packages/runtime/container-runtime/src/summaryManagerUtils.ts

```diff
@@ -0,0 +1,200 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IEvent } from "@fluidframework/common-definitions";
+import { assert, TypedEventEmitter } from "@fluidframework/common-utils";
+import { IQuorum, ISequencedClient } from "@fluidframework/protocol-definitions";
+
+export const summarizerClientType = "summarizer";
+
+export interface ITrackedClient {
+    readonly clientId: string;
+    readonly sequenceNumber: number;
+}
+
+interface ILinkedClient extends ITrackedClient {
+    prevClient: ILinkedClient | undefined;
+    nextClient: ILinkedClient | undefined;
+}
+
+interface IFullClient extends ILinkedClient {
+    readonly isSummarizer: boolean;
+}
+
+interface IOrderedClientsEvents extends IEvent {
+    (event: "currentChange", listener: (client: ITrackedClient | undefined) => void);
+    (event: "summarizerChange", listener: (summarizerCount: number) => void);
+}
+
+export class OrderedClients extends TypedEventEmitter<IOrderedClientsEvents> {
+    private readonly clientMap = new Map<string, IFullClient>();
+    private readonly rootClient: ILinkedClient = {
+        sequenceNumber: -1,
+        clientId: "",
+        prevClient: undefined,
+        nextClient: undefined,
+    };
+    private lastClient: ILinkedClient = this.rootClient;
+    private nonSummarizerCount = 0;
+    private summarizerCount = 0;
+    private currentClient: ILinkedClient | undefined;
+
+    constructor(quorum: Pick<IQuorum, "getMembers" | "on">) {
+        super();
+        const members = quorum.getMembers();
+        for (const [clientId, client] of members) {
+            this.addClient(clientId, client);
+        }
+
+        quorum.on("addMember", this.addClient);
+        quorum.on("removeMember", this.removeClient);
+
+        this.resetCurrentClient();
+    }
+
+    private readonly addClient = (clientId: string, client: ISequencedClient) => {
+        // Have to undefined-check client.details for backwards compatibility
+        const isSummarizer = client.client.details?.type === summarizerClientType;
+        const newClient: IFullClient = {
+            clientId,
+            sequenceNumber: client.sequenceNumber,
+            isSummarizer,
+            prevClient: undefined,
+            nextClient: undefined,
+        };
+        this.clientMap.set(clientId, newClient);
+
+        if (newClient.isSummarizer) {
+            this.summarizerCount++;
+            this.emit("summarizerChange", this.summarizerCount);
+            return;
+        }
+
+        // Normal case is adding the latest client, which will bypass loop.
```

Comment:
not normal being the initiall population of the list?

Reply:
Theoretically I guess, I'm not really sure what case allows them to come in unordered, since they are serialized as an array and already sorted naturally. I did not want to assume that serialization/deserialization rule just yet though.